### PR TITLE
More join types

### PR DESCRIPTION
**Pull request by max-hoffman**
Add `FullOuterJoin`, `SemiJoin`, and `AntiJoin`.

None of these new join nodes are safe for join ordering transformations yet. They are explicitly excluded from join planning.

The getField indexes for these three nodes' join conditions deserve more consideration. I excluded them from auto-fixup after manually correcting join condition get fields for the appropriate schemas.

`FullOuterJoin` uses a union distinct execution operator, which is correct but a lot slower than a merge join-esque operator.

`SemiJoin` and `AntiJoin` rearrange subquery expression scopes. I separate `resolve` and `finalizeSubqueryExpressions` to perform decorrelation before predicate pushdown (where we were panicking on FixUpExpressions) and join ordering (we want to decorrelate scopes before join planning).

Other:
- query plan tests added for exist hoisting edge cases i did not catch on first pass
- fixed bug with CTE stars

### CODE REVIEW

File: sql/analyzer/hoist_select_exists.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+			return plan.NewAntiJoin(left, right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
+		case plan.SemiJoinType:
+			return plan.NewSemiJoin(left, right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
+		default:
+			panic("expected SemiJoinType or AntiJoinType")
+		}
+	})
+}
+
+// pluckCorrelatedExistsSubquery scans a filter for [note] WHERE EXISTS, and then attempts to
```
Comment (zachmu) [inline sql/analyzer/hoist_select_exists.go:76]: what is `[note]`

File: sql/analyzer/hoist_select_exists.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+// extract the subquery, correlated filters, a modified outer scope (net subquery and filters),
+// and the new target joinType
+func pluckCorrelatedExistsSubquery(filter *plan.Filter) (sql.Node, sql.Node, []sql.Expression, plan.JoinType) {
```
Comment (zachmu) [inline sql/analyzer/hoist_select_exists.go:79]: This return type is a bit much, you should define a struct with named fields

File: sql/analyzer/hoist_select_exists.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		filters := splitConjunction(f.Expression)
+		var newFilters []sql.Expression
+		for _, f := range filters {
+			var outerRef bool
+			transform.InspectExpr(f, func(e sql.Expression) bool {
+				gf, ok := e.(*expression.GetField)
+				if !ok {
+					return false
+				}
+				if gf.Index() < scopeLen {
```
Comment (zachmu) [inline sql/analyzer/hoist_select_exists.go:136]: This isn't a scope len, as used

And in fact you aren't passing the scope from the top-level rule down into these functions that would need it

Why does this work? Are you missing a test case?

File: sql/analyzer/resolve_ctes.go [hunk 4]
```diff
@@ -142,11 +142,7 @@ func stripWith(ctx *sql.Context, a *Analyzer, scope *Scope, n sql.Node, ctes map
 		subquery := cte.Subquery
 
 		if len(cte.Columns) > 0 {
-			schemaLen := schemaLength(subquery)
-			if schemaLen != len(cte.Columns) {
-				return nil, sql.ErrColumnCountMismatch.New()
-			}
-
+			// wait until we've resolved stars to error for column mismatch
```
Comment (zachmu) [inline sql/analyzer/resolve_ctes.go:145]: This comment really only refers to code that was deleted

A better one might be `We don't validate the number of columns in the CTE schema until later, see (code ref)`

File: sql/plan/join.go [hunk 5]
```diff
@@ -27,13 +28,29 @@ import (
...
 	inMemoryJoinKey        = "INMEMORY_JOINS"
 	inMemoryJoinSessionVar = "inmemory_joins"
 )
 
 var useInMemoryJoins = shouldUseMemoryJoinsByEnv()
 
+type JoinType uint16
+
+const (
+	UnknownJoinType   JoinType = iota // UnknownJoinType
```
Comment (zachmu) [inline sql/plan/join.go:43]: nit: generally speaking we name this kind of enum constant with the type first, so that all the possible values sort next to eachother

File: sql/plan/join.go [hunk 6]
```diff
@@ -51,148 +68,175 @@ type JoinNode interface {
 }
 
 // joinStruct contains all the common data fields and implements the commom sql.Node getters for all join types.
```
Comment (zachmu) [inline sql/plan/join.go:70]: comment out of date

File: sql/plan/join.go [hunk 7]
```diff
@@ -51,148 +68,175 @@ type JoinNode interface {
...
-type joinStruct struct {
+type joinBase struct {
 	BinaryNode
-	Cond       sql.Expression
+	Filter     sql.Expression
```
Comment (zachmu) [inline sql/plan/join.go:73]: Prefer cond, is there a particular reason you changed this?

File: sql/plan/join.go [hunk 8]
```diff
@@ -51,148 +68,175 @@ type JoinNode interface {
...
 }
 
 // Schema implements the Node interface.
-func (j *InnerJoin) Schema() sql.Schema {
-	return append(j.left.Schema(), j.right.Schema()...)
+func (j *joinBase) Schema() sql.Schema {
+	panic("join implementation should provide schema")
+}
+
+func constructNewJoin(base *joinBase) JoinNode {
```
Comment (zachmu) [inline sql/plan/join.go:112]: Smart, I like it

File: sql/plan/join.go [hunk 9]
```diff
@@ -723,6 +619,135 @@ func (i *joinIter) Close(ctx *sql.Context) (err error) {
...
+
+func (i *existsIter) Next(ctx *sql.Context) (sql.Row, error) {
+	for {
+		r, err := i.primary.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		left := i.originalRow.Append(r)
+		rIter, err := i.secondaryProvider.RowIter(ctx, left)
+		defer rIter.Close(ctx)
```
Comment (zachmu) [inline sql/plan/join.go:671]: missing error handling on this close

**Review by zachmu — APPROVED** [review]
Mostly looks good, just a few comments

For future PRs, it's easier to review these when split into smaller functional chunks (although this wasn't egregious)

File: sql/analyzer/hoist_select_exists.go (hunk 1 above)
Comment (max-hoffman) [reply sql/analyzer/hoist_select_exists.go:76]: autocorrect i think, meant `[NOT] WHERE EXISTS`

File: sql/plan/join.go (hunk 7 above)
Comment (max-hoffman) [reply sql/plan/join.go:73]: filters are more mobile in the new join ordering algo, seems easier to just call them all filters. At some point i'd like to split these into a list of filters rather than an `expression.JoinAnd(filters...)`, probably memoized, index selection doesn't have to be as painful as it is right now.

File: sql/analyzer/hoist_select_exists.go (hunk 3 above)
Comment (max-hoffman) [reply sql/analyzer/hoist_select_exists.go:136]: good catch, subquery expression, added test

File: sql/plan/join.go (hunk 8 above)
Comment (max-hoffman) [reply sql/plan/join.go:113]: :)
